src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,302 @@
+                    release.filename = val.get_string ();
+                    break;
+                case "Name":
+                    release.name = val.get_string ();
+                    break;
+                case "Summary":
+                    release.summary = val.get_string ();
+                    break;
+                case "Version":
+                    release.version = val.get_string ();
+                    break;
+                case "Description":
+                    release.description = val.get_string ()
+                    .replace ("<p>", "")
+                    .replace ("</p>", "\n\n")
+                    .replace ("<li>", " • ")
+                    .replace ("</li>", "\n")
+                    .replace ("<ul>", "")
+                    .replace ("</ul>", "\n")
+                    .replace ("<ol>", "") // TODO: add support for ordered lists
+                    .replace ("</ol>", "\n")
+                    .strip ();
+                    break;
+                case "Protocol":
+                    release.protocol = val.get_string ();
+                    break;
+                case "RemoteId":
+                    release.remote_id = val.get_string ();
+                    break;
+                case "AppstreamId":
+                    release.appstream_id = val.get_string ();
+                    break;
+                case "Checksum":
+                    release.checksum = val.get_string ();
+                    break;
+                case "Vendor":
+                    release.vendor = val.get_string ();
+                    break;
+                case "Size":
+                    release.size = val.get_uint64 ();
+                    break;
+                case "License":
+                    release.license = val.get_string ();
+                    break;
+                case "TrustFlags":
+                    release.flag = (Fwupd.ReleaseFlag) val.get_uint64 ();
+                    break;
+                case "InstallDuration":
+                    release.install_duration = val.get_uint32 ();
+                    break;
+                case "Uri":
+                    release.uri = val.get_string ();
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async string? download_file (Fwupd.Device device, string uri) {
+        var path = get_path (uri);
+
+        File server_file = File.new_for_uri (uri);
+        File local_file = File.new_for_path (path);
+
+        bool result;
+        try {
+            result = yield server_file.copy_async (local_file, FileCopyFlags.OVERWRITE, Priority.DEFAULT, null, (current_num_bytes, total_num_bytes) => {});
```

Comment:
I think I'd like to see something done with this progress callback in the future so that people with slow connections or that are downloading large files know something is happening. I guess add a TODO comment for now and we'll open issues for all of them once this is merged.